updatehub/src/states/mod.rs

```diff
@@ -184,6 +190,33 @@ where
     }
 }
 
+fn handle_startup_callbacks(
+    settings: &Settings,
+    runtime_settings: &mut RuntimeSettings,
+) -> crate::Result<()> {
+    if let Some(expected_set) = runtime_settings.update.upgrade_to_installation {
+        info!("Booting from a recent installation");
+        if expected_set == firmware::installation_set::active()?.0 {
+            match firmware::validate_callback(&settings.firmware.metadata)? {
+                Transition::Cancel => {
+                    warn!("Validate callback has failed");
+                    firmware::installation_set::swap_active()?;
+                    warn!("Swapped active installation set and rebooting");
+                    easy_process::run("reboot")?;
+                }
+                Transition::Continue => firmware::installation_set::validate()?,
+            }
+            return Ok(());
+        } else {
+            warn!("Agent is not running on expected installation set, running rollback callback");
+            firmware::rollback_callback(&settings.firmware.metadata)?;
```

Comment:
Here we need to handle the error case, right?

I wonder if the callback somehow fails, what the agent does?

Reply:
Failing to run the `handle_startup_callbacks` as hole is handled by the `run` function:
```Rust
    if let Err(e) = handle_startup_callbacks(&settings, &mut runtime_settings) {
        error!("Failed to handle startup callbacks: {}", e);
    }
```